File: 07/d07-1.py
# Note: this solution assumes, that each directory is visited only once and the files in each
# directory are listed only once.
#

limit = 100000

paths = [""]
sizes = {}
for line in open("input.dat", "r"):
    if not line.strip():
        break
    if line.startswith("$ cd "):
        newpath = line[5:-1]
        if newpath == "/":
            paths = [paths[0]]
        elif newpath == "..":
            paths.pop(-1)
        else:
            paths.append(newpath)
    elif not line.startswith("dir") and not line.startswith("$ ls"):
        word0, word1 = line.split(maxsplit=2)
        name, size = word1, int(word0)
        for ipath in range(len(paths)):
            path = "/".join(paths[0 : ipath + 1])
            sizes[path] = sizes.get(path, 0) + size

summed_sizes = sum([value for value in sizes.values() if value <= limit])
print(f"Summed directory sizes: {summed_sizes}")

File: 07/d07-1.f90
! Note: this solution assumes, that each directory is visited only once and the files in each
! directory are listed only once.
!
program d07_1
  implicit none

  integer, parameter :: size_limit = 100000

  integer, parameter :: max_line_len = 100

  integer, allocatable :: dirsizes(:), pathinds(:)
  integer :: fd, iostat, filesize, dirsizesum
  character(max_line_len) :: buffer, filename

  dirsizes = [0]
  pathinds =  [1]
  open(newunit=fd, file="input.dat", action="read")
  do
    read(fd, "(a)", iostat=iostat) buffer
    if (iostat /= 0 .or. len_trim(buffer) < 4) exit
    if (buffer(1:4) == "$ cd") then
      select case (trim(buffer(6:)))
      case ("..")
        pathinds = pathinds(1 : size(pathinds) - 1)
      case ("/")
        pathinds = [pathinds(1 : 1)]
      case default
        dirsizes = [dirsizes, 0]
        pathinds = [pathinds, size(dirsizes)]
      end select
    else if (buffer(1:4) /= "dir " .and. buffer(1:4) /= "$ ls") then
      read(buffer, *) filesize, filename
      dirsizes(pathinds) = dirsizes(pathinds) + filesize
    end if
  end do
  close(fd)

  dirsizesum = sum(pack(dirsizes, dirsizes <= size_limit))
  print "(a, i0)", "Minimal directory size to delete: ", dirsizesum

end program d07_1
